Sources/StreamChatSwiftUI/ChatChannel/Composer/MessageComposerViewModel.swift

```diff
@@ -124,9 +130,40 @@ open class MessageComposerViewModel: ObservableObject {
     @Published public var suggestions = [String: Any]()
     @Published public var cooldownDuration: Int = 0
     @Published public var attachmentSizeExceeded: Bool = false
+    @Published public var recordingState: RecordingState = .initial {
+        didSet {
+            if case let .recording(location) = recordingState {
+                if location.y < RecordingConstants.lockMaxDistance {
+                    recordingState = .locked
+                } else if location.x < RecordingConstants.cancelMaxDistance {
+                    audioRecordingInfo = .initial
+                    recordingState = .initial
+                    stopRecording()
+                }
+            } else if recordingState == .showingTip {
+                DispatchQueue.main.asyncAfter(deadline: .now() + 2.0) { [weak self] in
+                    self?.recordingState = .initial
+                }
+            }
+        }
+    }
+    
+    @Published public var audioRecordingInfo = AudioRecordingInfo.initial
     
     public let channelController: ChatChannelController
     public var messageController: ChatMessageController?
+    public var waveformTargetSamples: Int = 100
+    public internal(set) var pendingAudioRecording: AddedVoiceRecording?
+    
+    internal lazy var audioRecorder: AudioRecording = {
+        let audioRecorder = StreamAudioRecorder()
```

Comment:
Is there a way for integrators to changes this with their own implementation? (Something similar to what Picslo needed a few days ago)